### PR TITLE
Yazan/model api implementation

### PR DESCRIPTION
**Pull request by santosjem**
We refactored the fields, implemented story and realm local client model methods and objects. We also added the constants to replace core strings.

### CODE REVIEW

File: app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ * Constants used as column values to be used throughout app
+ *
+ * @author  Yazan Kittaneh
+ */
+
+public class CONSTANTS {
+    public static final String PUBLICATIONID ="publicationId";
+    public static final String LASTUPDATED ="lastUpdated";
+    public static final String STORYID ="storyId";
+
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java]: unnecessary Whitespace

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java [hunk 2]
```diff
@@ -12,95 +12,133 @@
...
  */
 public class RealmStory extends RealmObject implements Story{
-    private RealmPublication publication;
-    private String publicationDate;
-    private String lastUpdated;
-    private RealmList<RealmAuthor> author;
-    private RealmReactions reactions;
-    private String fullText;
-    private String blurb;
+    private String datePublished;
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: All of these member variables should follow the "mVariable" name scheme. 
i.e: mDatePublished

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java [hunk 3]
```diff
@@ -12,95 +12,133 @@
...
+    private RealmList authors;
 
 
     /* Default constructor for Realm */
     public RealmStory(){}
 
-    public RealmStory(RealmPublication publication, String publicationDate, String lastUpdated,
-                      RealmList<RealmAuthor> author, RealmReactions reactions, String fullText,
-                      String blurb, String title, String thumbnailUrl) {
+    public RealmStory(String datePublished, String brief, String headerImage, String publication,
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: Can you make this chunk more readable? Having all the inputs aligned. 
i.e:
String datePublished, 
String brief,
...

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java [hunk 4]
```diff
@@ -12,95 +12,133 @@
...
     @Override
-    public String getPublicationDate() {
-        return this.publicationDate;
+    public String getDatePublished() {
+        return this.datePublished;
     }
 
     @Override
-    public String getLastUpdated(){ return this.lastUpdated;}
+    public String getDateEdited() { return this.dateEdited; }
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: keep code consistent. 
either:  
public String getDateEdited() { return this.dateEdited; }
or:
 public String getDateEdited() { 
    return this.dateEdited; 
}

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 5]
```diff
@@ -34,12 +40,28 @@ public void addToSubscribedPublications(int publicationId) {
 
     @Override
     public void savePublications(List<Publication> publications) {
+        Realm realm = Realm.getDefaultInstance();
+        realm.beginTransaction();
 
+        for(int i = 0; i < publications.size(); i++) {
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: consider doing a foreach loop to be more readable

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 6]
```diff
@@ -49,22 +71,43 @@ public void savePublication(Publication publication) {
 
     @Override
     public List<Publication> getAllPublications() {
-        return null;
+        Realm realm = Realm.getDefaultInstance();
+        RealmResults<RealmPublication> pubResults = realm.where(RealmPublication.class).findAll();
+        List<Publication> mResults = new ArrayList<>();
+        mResults.addAll(pubResults.subList(0, pubResults.size()));
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: Are we putting Publication objects in the List<> or are we putting RealmObjects?

File: app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package edu.grinnell.grinnell_publications_android;
+
+/**
+ * Constants used as column values to be used throughout app
+ *
+ * @author  Yazan Kittaneh
+ */
+
+public class CONSTANTS {
+    public static final String PUBLICATIONID ="publicationId";
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java]: should be in the form PUBLICATION_ID, LAST_UPDATED

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 8]
```diff
@@ -34,12 +40,28 @@ public void addToSubscribedPublications(int publicationId) {
...
     @Override
     public void savePublication(Publication publication) {
+        Realm realm = Realm.getDefaultInstance();
+        realm.beginTransaction();
 
+        RealmResults<RealmPublication> pubList = realm.where(RealmPublication.class)
+                .equalTo(CONSTANTS.PUBLICATIONID, publication.getPublicationId())
+                .findAll();
+
+        if(pubList.size() == 0) {
```
Comment (YazanKittaneh) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: You should comment on what this code is doing. It's unfamiliar to whoever doesn't know how Realm works.

**Review by YazanKittaneh — CHANGES_REQUESTED** [review]
Check out the changes! (Some of it is on my code but you should check it anyway ;) )

File: app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java (hunk 7 above)
Comment (LarryAsante) [reply app/src/main/java/edu/grinnell/grinnell_publications_android/CONSTANTS.java]: +1 to this!

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java (hunk 2 above)
Comment (LarryAsante) [reply app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: +1 to this too! Yep, let's stick to the conventional member variable naming scheme! Thanks for catching that!

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java [hunk 9]
```diff
@@ -12,95 +12,133 @@
...
         this.title = title;
-        this.thumbnailUrl = thumbnailUrl;
+        this.content = content;
+        this.authors = authors;
     }
 
     /** Setters */
-    public void setPublication(RealmPublication publication){
+    public void setPublication(String publication) {
+        if (publication.isEmpty() || publication == null) {
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: Nice defensive checks here!

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java (hunk 3 above)
Comment (LarryAsante) [reply app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: The method should be formatted properly. Also, we should definitely look into *not* having such a long constructor.

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java [hunk 10]
```diff
@@ -12,95 +12,133 @@
...
+            return;
+        }
         this.publication = publication;
     }
-    public void setPublicationDate(String publicationDate){
-        this.publicationDate = publicationDate;
+
+    public void setDatePublished(String datePublished) {
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Models/Realm/RealmStory.java]: Please document all methods you make public so the client who calls this method has an easier time knowing what it does, what the parameter are and what they mean. When someone calls this method, they can just hover over it and see the documentation rather than having to go to 

A good example is @as-f's documentation of his public methods here: https://github.com/GrinnellAppDev/DB-Android/pull/3/files#diff-23c4fc42ab674199182b5a617e912235

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/NetworkClient.java [hunk 11]
```diff
@@ -60,4 +60,7 @@ public Story getFullStoryById(int storyId) {
     public User getProfile() {
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/NetworkClient.java:60]: Please don't return null! That's dangerous! If you return null and the client tries to access the return value of this method, we'll crash woefully with a `NullPointerException`

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/NetworkClient.java [hunk 12]
```diff
@@ -60,4 +60,7 @@ public Story getFullStoryById(int storyId) {
...
         return null;
     }
+
+    @Override
+    public Publication getPublicationById(String publicationId) { return null;}
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/NetworkClient.java:65]: See comment above

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java (hunk 6 above)
Comment (LarryAsante) [reply app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: +1 to that. Also, what is this "0" magic number here? Make a constant with a descriptive variable name.

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 13]
```diff
@@ -49,22 +71,43 @@ public void savePublication(Publication publication) {
...
     @Override
     public List<Story> getRecentStories(List<Integer> subscribedPublicationIds, Date mostRecentStory) {
-        return null;
+        Realm realm = Realm.getDefaultInstance();
+        List<Story> allStories = new ArrayList<>();
+
+        for (Integer currentPublicationId: subscribedPublicationIds) { //for each subscribed publication
+            RealmResults<RealmStory> publicationStories = realm
+                    .where(RealmStory.class)
+                    .equalTo(CONSTANTS.PUBLICATIONID, currentPublicationId)
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: You should use static imports to shorten this code so you're just calling `PUBLICATION_ID`

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 14]
```diff
@@ -49,22 +71,43 @@ public void savePublication(Publication publication) {
...
+                    .greaterThan(CONSTANTS.LASTUPDATED, mostRecentStory)
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: See comment above

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 15]
```diff
@@ -49,22 +71,43 @@ public void savePublication(Publication publication) {
...
+        }
+        //TODO: sort all stories by date published
+        return allStories;
     }
 
     @Override
     public Story getFullStoryById(int storyId) {
-        return null;
+        Realm realm = Realm.getDefaultInstance();
+        return realm.where(RealmStory.class).equalTo(CONSTANTS.STORYID, storyId).findFirst();
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: See comment above

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 16]
```diff
@@ -74,16 +117,30 @@ public Story getFullStoryById(int storyId) {
 
     @Override
     public List<Story> getRecentStories(int seriesId, Date mostRecentStoryInSeries) {
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: What happens if the date is null? Let's handle such edge cases graciously.

File: app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java [hunk 17]
```diff
@@ -74,16 +117,30 @@ public Story getFullStoryById(int storyId) {
...
 
     @Override
     public List<Story> getBookmarkedStories() {
-        return null;
+        return new ArrayList<>();
     }
 
     @Override
     public User getProfile() {
         return null;
```
Comment (LarryAsante) [inline app/src/main/java/edu/grinnell/grinnell_publications_android/Services/Implementation/RealmLocalClient.java]: Nope, please don't return null

**Review by LarryAsante — CHANGES_REQUESTED** [review]
Thanks for the work done here! Let's make the necessary changes and we'll be good to go!
